Implement typed dictionaries

Repiteo opened the pull request:
This aims to add the ability to bind typed dictionaries to script & GDScript. The implementation takes heavy inspiration from the existing typed array format. ~~The primary difference is the ability to specify type for *just* a key, leaving the value as typeless much like current dictionaries.~~ Both key and value need their types specified for typed dictionaries, with `Variant` allowing for a typeless equivalent.

### Syntax
```gdscript
var typed_key_value: Dictionary[int, String] = { 1: "first value", 2: "second value", 3: "etc" }
var typed_key: Dictionary[int, Variant] = { 0: "any value", 10: 3.14, 100: null }
var typed_value: Dictionary[Variant, int] = { "any value": 0, 123: 456, null: -1 }
```

### Editor
![23-06-26 10-02-25 godot windows editor dev x86_64 san](https://github.com/godotengine/godot/assets/17489292/81ae9f94-7af4-4813-8715-3766c2aac084)

### Issues
**All core issues are resolved!**
~~There are a fair number of areas that still require fixes, in addition to whatever lingering bugs/errors there are. In particular:~~
- [x] No way currently to make a typeless key and a typed value
- [x] `Variant` text appears as white instead of green in GDSCript if used as the value:
![23-06-26 10-18-36 godot windows editor dev x86_64 san](https://github.com/godotengine/godot/assets/17489292/06d0f1c1-afa8-4e8a-96b3-0cb588380af9)
- [x] Editor fails to update creation key/value sections if changing type
- [x] Editor incorrectly perceives the initial typed key/value as null
- [x] Editor fails to setup exported dictionaries during runtime
- [x] Editor won't recognize a typed key passed with the incorrect type as invalid in a GDScript function (that is, when using the `[] operator`); it throws an error if attempting to do so in runtime & does recognize invalid keys when creating a dictionary initially
- [x] Editor will treat any Object variant as a "Resource" regardless of provided type
- [x] Dictionary fails to recognize local script types in curly-bracket initialization during runtime
- [x] Potential memory leak with `container_element_type` being a list now instead of a single item. (This is inexperience with the engine on my end, dunno if it needs something special to memdelete array entries)
- [x] Implementation of GDScript tests

Closes godotengine/godot-proposals#56

Repiteo committed:
Implement typed dictionaries

lufog commented:
I would prefer:
```
var my_dictionary: Dictionary[int, Variant]
var my_dictionary: Dictionary[Variant, int]
```

Just `Dictionary[int]` a little confusing.

dalexeev commented on core/object/object.h:
```diff
@@ -80,6 +80,7 @@ enum PropertyHint {
 	PROPERTY_HINT_INT_IS_OBJECTID,
 	PROPERTY_HINT_INT_IS_POINTER,
 	PROPERTY_HINT_ARRAY_TYPE,
+	PROPERTY_HINT_DICTIONARY_TYPE,
```
I think it's better to add the hint to the end of the enum, without shifting the values.

Repiteo commented:
@lufog Great point! I adjusted the code to handle exactly that, but it revealed some other pitfalls in the process (mainly the process of transfering container types from one DataType to another) so I'll have to iron that out before showing off a new version

In the meantime I've added documentation descriptions & made it possible to use a constructor to make a typed dictionary in GDScript, much like typed arrays

Repiteo commented:
GDScript tests implemented! They're largely lifted from the tests given to typed Arrays so more specialization will be required, but they're a solid starting point to see what's awry off the bat

Namely, typed arrays seem to be instantiating with null keys & values, though they're recognized as typed. I'm not quite sure yet where the incorrect implementation is happening, but that's my current focus as it very well may be the linchpin of the editor errors

Bromeon commented:
With typed arrays, we have the variance problem:
```gdscript
var typed: Array[int] = [1, 2, 3, 4]
var as_untyped: Array = typed

as_untyped[0] = "string" # silently fails at runtime
```

I guess the same problem exists here, but for both keys and values?
How do we deal with that?

dalexeev commented:
> With typed arrays, we have the variance problem

The exception for `Array` (aka `Array[Variant]`) was intentionally added as a convenience to users of untyped GDScript.

```gdscript
var a: Array[Node2D]
var b: Array[Node] = a # Error
var c: Array[Variant] = a # Unsafe, but ok.
var d: Array = a # Unsafe, but ok.
```

![](https://github.com/godotengine/godot/assets/47700418/0349887a-429d-4709-8eee-b7da13fac45c)

Bromeon commented:
> The exception for `Array` (aka `Array[Variant]`) was intentionally added as a convenience to users of untyped GDScript.

Yes, but it's still a hole in the type system. I understand the desire to pass typed arrays to APIs accepting untyped ones, and for _reading_, this is completely sound. The covariance violation occurs when _writing_ to an `Array` that points to a `Array[T]`. This cannot be checked at parse time, it's a runtime check (and IIRC it's silent do-nothing, no error at the moment).

My question is, is the same covariance conversion ("upcast") planned for dictionaries on each of the generic parameters? That is:
* `Dictionary[int, String]` -> `Dictionary[int, Variant]`
* `Dictionary[int, String]` -> `Dictionary[Variant, String]`
* `Dictionary[int, Variant]` -> `Dictionary[Variant, Variant]` == `Dictionary`

And if yes, are we fully aware of the pitfalls?

dalexeev commented:
> This cannot be checked at parse time, it's a runtime check (and IIRC it's silent do-nothing, no error at the moment).

![](https://github.com/godotengine/godot/assets/47700418/aded01e3-49dd-4f97-bb69-834ff3b74e06)

However, this is not directly related to the PR. PRs of this kind tend to garner large numbers of comments that are difficult to navigate. If you'd like to discuss this, I suggest the `#gdscript` channel on Rocket Chat.

Repiteo commented:
The holes in the type system certainly have been a bit of a headache to navigate, but attempting to overhaul that system as a whole is probably a bit beyond this PR (at least as I initially envisioned it). This is more about setting up the system that currently exists to dictionaries, warts and all. If the current implementation were to somehow inhibit the ability to implement typed dictionaries in the first place, _then_ changes to the system would make sense

On that note, I've figured out why variants were appearing as null & not changing when swapping types, so both those fixes are in place. As it stands, the two biggest remaining issues are:
- The editor treats any Object variant as "Resource", regardless of the provided type (need to figure out how typed arrays handle this)
- A dictionary fails to properly parse provided key/value if there's unorthodox syntax (eg: [`typed_dictionary_assign_wrong_to_typed.out`](https://github.com/godotengine/godot/pull/78656/files#diff-40bda9d223931c197707b7c9871aa42224610cc7ce0971e71a496c989ff20a92) has the wrong error, thinks an entirely different type is passed on as the value (from reading the key?)

Repiteo commented:
This is nearing the end of the draft stage! Clearing out most other straggling bugs clarified some of my prior listed issues & made me realize they were entirely seperate problems altogether. As it stands, there's only one major issue remaining: recognizing local script types in curly-bracket initialization during runtime

Currently `VariantInternal::get_dictionary()` retrieves a dictionary as expected, excluding the script types. Once I figure out why this process works without a hitch for typed arrays & can apply it to typed dictionaries, I'll be confident in submitting for review ⭐

(Sidenote, I have no idea why the Linux doubles/sanitizer build sometimes fails. It feels like it's a tossup whether it succeeds or not, and some advice on how to replicate that test on a Windows 11 environment would be very appreciated!)

Repiteo commented:
That took longer than expected (there's gotta be a better way of debugging OPCodes…), but I finally got the last major roadblock figured out. Everything is now in order & ready for review!

Repiteo commented:
Fixed a regression causing objects to not be selectable from the Editor if the Dictionary was untyped (or rather, they could be clicked, but it'd just make it nil instead). Alongside that was a fix for the objects more broadly, where an uninitialized editor key would wrongly pull its type from the value type

Outside of that, I went ahead and readded the original implementation idea of typing just the key as an option in GDScript (ie: `Dictionary[int]` as an equivalent for `Dictionary[int, Variant]`). After seeing Python have the functionality to do the same, it felt right at home in this environment

Vennnot commented:
Can't wait for this in 4.2!

adamscott commented:
@Vennnot I know that you're really enthusiast about the perspective of typed dictionaries, but please, don't comment only to state this.
- Every contributor gets a ping for essentially nothing;
- [With the new release management policy](https://godotengine.org/article/release-management-4-1/), we have a shorter dev cycle for each version, so for many reasons, it is possible that this feature would not be merged for 4.2;
- There's already a way to mark enthusiasm for the PR: the reactions.

Thank you.

Maran23 commented:
There are merge conflicts, so this branch needs a rebase. :)

AThousandShips commented on core/variant/dictionary.cpp:
```diff
@@ -313,6 +351,146 @@ Array Dictionary::values() const {
+
+	if ((typed_key == typed_key_source || typed_key.type == Variant::NIL || (typed_key_source.type == Variant::OBJECT && typed_key.can_reference(typed_key_source))) &&
+			(typed_value == typed_value_source || typed_value.type == Variant::NIL || (typed_value_source.type == Variant::OBJECT && typed_value.can_reference(typed_value_source)))) {
+		// from same to same or
+		// from anything to variants or
+		// from subclasses to base classes
```
```suggestion
		// From same to same or,
		// from anything to variants or,
		// from subclasses to base classes.
```
[Comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide) nitpick

AThousandShips commented on core/variant/dictionary.cpp:
```diff
@@ -313,6 +351,146 @@ Array Dictionary::values() const {
+	Variant *value_data = value_array.ptrw();
+
+	if (typed_key == typed_key_source || typed_key.type == Variant::NIL || (typed_key_source.type == Variant::OBJECT && typed_key.can_reference(typed_key_source))) {
+		// from same to same or
+		// from anything to variants or
+		// from subclasses to base classes
```
```suggestion
		// From same to same or,
		// from anything to variants or,
		// from subclasses to base classes.
```

AThousandShips commented on core/variant/dictionary.cpp:
```diff
@@ -313,6 +351,146 @@ Array Dictionary::values() const {
+	}
+
+	if (typed_value == typed_value_source || typed_value.type == Variant::NIL || (typed_value_source.type == Variant::OBJECT && typed_value.can_reference(typed_value_source))) {
+		// from same to same or
+		// from anything to variants or
+		// from subclasses to base classes
```
```suggestion
		// From same to same or,
		// from anything to variants or,
		// from subclasses to base classes.
```

AThousandShips commented on core/variant/dictionary.cpp:
```diff
@@ -313,6 +351,146 @@ Array Dictionary::values() const {
+			ERR_FAIL_COND_MSG(ce.error, vformat(R"(Unable to convert value at key "%s" from "%s" to "%s".)", key_data[i - 1], Variant::get_type_name(value->get_type()), Variant::get_type_name(typed_value.type)));
+		}
+	} else if (Variant::can_convert_strict(typed_value_source.type, typed_value.type)) {
+		// from primitives to different convertible primitives
```
```suggestion
		// From primitives to different convertible primitives.
```

AThousandShips commented on core/variant/typed_dictionary.h:
```diff
@@ -0,0 +1,1273 @@
+	}
+};
+
+//specialization for the rest of variant types
```
```suggestion
// Specialization for the rest of variant types.
```

AThousandShips commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -752,9 +776,10 @@ void EditorPropertyDictionary::_change_type(Object *p_button, int p_index) {
-	// Do not allow nil as valid key. I experienced errors with this
-	if (object->get_new_item_key().get_type() == Variant::NIL) {
-		return;
+	// Treat empty object key as nil to prevent editor-only duplicates
```
```suggestion
	// Treat empty object key as nil to prevent editor-only duplicates.
```

AThousandShips commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2500,8 +2580,14 @@ void GDScriptAnalyzer::reduce_assignment(GDScriptParser::AssignmentNode *p_assig
+		update_array_literal_element_type(static_cast<GDScriptParser::ArrayNode *>(p_assignment->assigned_value), assignee_type.get_container_element_type(0));
+	}
+
+	// Apply similar check but for dictionaries
```
```suggestion
	// Apply similar check but for dictionaries.
```

AThousandShips commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2579,8 +2665,8 @@ void GDScriptAnalyzer::reduce_assignment(GDScriptParser::AssignmentNode *p_assig
-			} else if (assignee_type.has_container_element_type() && !op_type.has_container_element_type()) {
-				// typed array assignee and untyped array result
+			} else if ((assignee_type.has_container_element_type(0) && !op_type.has_container_element_type(0)) || (assignee_type.has_container_element_type(1) && !op_type.has_container_element_type(1))) {
+				// typed assignee and untyped result
```
```suggestion
				// Typed assignee and untyped result.
```

AThousandShips commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -3945,15 +3952,28 @@ bool GDScriptParser::export_annotations(const AnnotationNode *p_annotation, Node
+		if (export_type.builtin_type == Variant::DICTIONARY && (export_type.has_container_element_type(0) || export_type.has_container_element_type(1))) {
+			DataType inner_type = export_type.get_container_element_type_or_default(1);
+			export_type = export_type.get_container_element_type_or_default(0); // Use inner types for, start with key.
+			export_type.set_container_element_type(0, inner_type); // Store earlier extracted value within key to separately parse after
```
```suggestion
			export_type.set_container_element_type(0, inner_type); // Store earlier extracted value within key to separately parse after.
```

AThousandShips commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -3945,15 +3952,28 @@ bool GDScriptParser::export_annotations(const AnnotationNode *p_annotation, Node
-			push_error(R"(Cannot use simple "@export" annotation because the type of the initialized value can't be inferred.)", p_annotation);
-			return false;
+			if (is_dict) {
+				// Dictionary allowed to have a variant key/value
```
```suggestion
				// Dictionary allowed to have a variant key/value.
```

AThousandShips commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -4044,6 +4064,109 @@ bool GDScriptParser::export_annotations(const AnnotationNode *p_annotation, Node
+			}
+			key_prefix += ":" + variable->export_info.hint_string;
+
+			// Now parse value
```
```suggestion
			// Now parse value.
```

Maran23 commented:
Friendly reminder: Has another merge conflict 😮 
I plan to test this today :)

Maran23 commented:
So I found one thing which seems like it is not supported (yet?): Nested typed collections.
When I do the following:
![image](https://github.com/godotengine/godot/assets/66004280/d2e22628-cff0-4e4a-b395-ac356f41c09b)
I get the following error:
![image](https://github.com/godotengine/godot/assets/66004280/8d1c9b00-769f-4816-81e9-2d08e04ea0c3)
Which is a bit unfortunate as I mostly have a case where I map something to an `Array` of something else.

Otherwise everything works.
I use the `Dictionary` a lot for item lookups, e.g. something like `Dictionary[String, ItemInfo]` or `Dictionary[Item, int]` and no problems with that (`Item` and also `ItemInfo` are `Resource` objects in this case).

Repiteo commented:
@Maran23 Nested typed collections are something that isn't supported as a whole; Arrays also lack this functionality

I'd love to eventually add nesting support, but that's something that'll have to be applied to the core system & I can't speak for how easily that can be done

Maran23 commented:
> @Maran23 Nested typed collections are something that isn't supported as a whole; Arrays also lack this functionality
> 
> I'd love to eventually add nesting support, but that's something that'll have to be applied to the core system & I can't speak for how easily that can be done

No problem! Just want to mention it for other people which want to test this change. :)

produno commented:
Lets hope this makes it into v4.2 before the feature freeze in a couple of weeks! Would be nice to have typed for loops and typed dictionaries complete. I think all we have left then is the potential to have nested types for Dicts and Arrays right?

AThousandShips commented:
This isn't likely to make it into 4.2, we have feature freeze in a week or so, and this is too big a feature to get rushed in so close to that

AThousandShips commented on modules/gdscript/gdscript_function.h:
```diff
@@ -147,24 +183,42 @@ class GDScriptDataType {
-		return *container_element_type;
+	GDScriptDataType get_container_element_type(int index) const {
+		ERR_FAIL_COND_V(container_element_type_size() <= index, GDScriptDataType());
+		ERR_FAIL_COND_V(container_element_types[index] == nullptr, GDScriptDataType());
```
```suggestion
		ERR_FAIL_NULL_V(container_element_types[p_index], GDScriptDataType());
```

AThousandShips commented on modules/gdscript/gdscript_parser.h:
```diff
@@ -151,37 +151,60 @@ class GDScriptParser {
-			return container_element_type != nullptr;
+		_FORCE_INLINE_ DataType get_container_element_type(int index) const {
+			ERR_FAIL_COND_V(container_element_type_size() <= index, DataType());
+			ERR_FAIL_COND_V(container_element_types[index] == nullptr, DataType());
```
```suggestion
			ERR_FAIL_NULL_V(container_element_types[pindex], DataType());
```

Bromeon commented on core/extension/gdextension_interface.h:
```diff
@@ -2109,6 +2109,33 @@ typedef GDExtensionVariantPtr (*GDExtensionInterfaceDictionaryOperatorIndex)(GDE
  */
 typedef GDExtensionVariantPtr (*GDExtensionInterfaceDictionaryOperatorIndexConst)(GDExtensionConstTypePtr p_self, GDExtensionConstVariantPtr p_key);
 
+/**
+ * @name dictionary_ref
+ * @since 4.2
+ *
+ * Sets a Dictionary to be a reference to another Dictionary object.
+ *
+ * @param p_self A pointer to the Dictionary object to update.
+ * @param p_from A pointer to the Dictionary object to reference.
+ */
+typedef void (*GDExtensionInterfaceDictionaryRef)(GDExtensionTypePtr p_self, GDExtensionConstTypePtr p_from);
```
Could you elaborate (in comment) when this function is supposed to be used by GDExtension? Especially in contrast to the copy constructor.

(I understand it exists for arrays as well, but might also benefit from some docs there).

Bromeon replied:
Also needs an update of `@since`, as this feature won't make it to 4.2.

Repiteo replied:
…You know, the more I look into it, the less I'm certain. It was initially a matter of achieving parity with the typed arrays so I just didn't question it, but I don't think this explicit setup is all that necessary. I'll probably hammer out a PR to get a feel for if it's needed

Repiteo replied:
Removed under the current assumption that `Dictionary::Operator=` will be able to replace this (see [this comment](https://github.com/godotengine/godot/pull/82497#issuecomment-1742150126))

Repiteo commented:
As this is safely out of consideration for 4.2, I've been expanding functionality once more. Namely: ensuring that documentation works properly in all areas & sweeping through existing Dictionary functions to make sure they work in a typed context

Right now it's throwing a fit about API changes, which is fair, but I'm not sure of the best way to resolve that yet. Probably going to spend the day looking into that and/or implementing typed dictionary editor support for C#

DaloLorn commented:
Is there any hope of it being ready by 4.*3*, though? I'm getting the distinct impression that I'll need this (especially the inspector upgrades!) to untangle the Gordian knot I've tied around my game data... this, and/or an implementation of godotengine/godot-proposals#4089... 😩

Repiteo commented:
@DaloLorn The scope of this PR is quite large, so there's no guarantee that it'll be incorporated at that time. To help mitigate that though, I'll be segmenting parts of this PR in order to make the core change, the typed dictionaries themselves, as painless to review/implement as possible. So far there's only #81662, but I'll have plenty of time for more during this feature-freeze

And while I'm here, I'll go over a few changes since last time:
- Node types were an area I didn't even realize needed a special NodePath conversion before, but that's now implemented in the same vein as typed arrays for packed scenes
- I've figured out where to put the API changes so it passes tests. The only exception is the `godot-cpp` check will always fail because it doesn't recognize typed dictionaries outright, but there's a equivalent PR for that repo anyway so it's a moot point
- Thanks to raulsantos in RocketChat, I now know where to look for C# inspector implementation & the accompanying sample project to debug in. This is pretty far removed from my usual debugging environment for the rest of the project, so I'm hoping to tap into my cursed Unity knowledge to get that fully functional sometime today

DaloLorn commented:
Honestly, as far as I'm concerned, even simply being able to hint to the inspector that an exported dictionary should be typed would be enough. Compile-time checking (to speak nothing of runtime checking!) is just icing on the cake in my book, and I wouldn't even be married to the notion of it surviving an inappropriately typed key-value pair. 😂

It's not a firm no, though, so there's still hope!

emilekm2142 commented:
Plz make this a priority. GDscript seems like a university project without this feature. Make the language usable.

AThousandShips commented:
Please don't bump issues or PRs without contributing significant new information. Please comment on the proposal instead

KoBeWi commented:
Shouldn't the node changes be in a separate PR, or at least a separate commit?
TypedArray was gradually introduced to the codebase (although replacing Arrays involved [much](https://github.com/godotengine/godot/pull/63959) [more](https://github.com/godotengine/godot/pull/64009) [changes](https://github.com/godotengine/godot/pull/64082)).

AThousandShips commented:
I'd say to introduce the feature in its own PR, to be merged individually, to test it and evaluate it, before applying it to core

Repiteo commented:
That's fair, the core conversions can wait. I do want to make sure they at least function in a bindable context, that was the main reason I was making those adjustments in the first place, but the changes *themselves* don't need to be handled in this specific PR

AThousandShips commented:
Also the changes to built in methods needs to be handled from a compatibility perspective, which needs to be discussed

Jakepc007 commented:
Would it be in the scope (or goal) of this PR to add type information to existing dictionaries? Such as `intersect_ray`?

```
var query := PhysicsRayQueryParameters2D.create(
Vector2.ZERO, 
Vector2.RIGHT
)
var result := space_state.intersect_ray(query)

func handle_result(result: 🤔):
  if result.collider:
    pass
```

I would imagine this would have some overlap with #76843 if such is the case. As those properties aren't even guaranteed.

AThousandShips commented:
It would break compatibility to do so imo so not guaranteed

YuriSizov commented:
> Would it be in the scope (or goal) of this PR to add type information to existing dictionaries?

It wouldn't be in scope for this PR, but it can be discussed as a follow-up to it.

IntangibleMatter commented:
> Would it be in the scope (or goal) of this PR to add type information to existing dictionaries? Such as `intersect_ray`?
> 
> ```
> var some_func():
>   var result := space_state.intersect_ray(query)
>   handle_result(result)
> 
> func handle_result(result: 🤔):
>   if result.collider:
>     pass
> ```
> 
> I would imagine this would have some overlap with #76843 if such is the case. As those properties aren't even guaranteed.

For Dictionaries with a precise type schema, but with multiple different types, a struct would likely make more sense. godotengine/godot-proposals#7903

DaloLorn commented:
Looks like the C# editor is failing to build?

IntangibleMatter commented:
Looks like all tests are passing... Can someone review this ASAP? It'd be awesome to have in 4.3, afaik feature freeze hasn't hit yet

Bromeon commented:
There are still open variance problems with `Array[T]` for which no one has a solution, see [my response here](https://github.com/godotengine/godot/pull/78656#issuecomment-1613075026) and [problem elaboration here](https://github.com/godotengine/godot-proposals/discussions/7364#discussioncomment-6576975). I'm not sure if it's wise to rush an implementation as long as we don't have a proper strategy for this (unless people agree that this is not an issue worth addressing).

DaloLorn commented:
I'd say "not worth addressing". I'm biased, though: I'm *positive* I'm not going to have trouble with it, and I *desperately* want to be able to tell the inspector what kind of object I want to put into my dictionaries.

(The proposal this PR is addressing is one of several prospective language upgrades I've been watching with great interest because they eliminate a lot of unnecessary busywork and complexity from my work on Fleet Ops.)

IntangibleMatter commented:
I agree with @DaloLorn . This feels like the kind of thing that's just... part of how the language works, and if/when it does get addressed, it'd make more sense to do at the same time as Arrays. There are *exponentially* more problems solved short-term and long-term than if we wait for every possible issue to be resolved. The covariance thing feels really nitpicky, and not like the kind of think 90% of users will worry about.

IMO it should be pushed as it is, Covariance is its own issue that needs to be resolved in its own discussion. This is just about getting statically typed dictionaries to *finally* be in the engine

ajreckof commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -776,9 +800,10 @@ void EditorPropertyDictionary::_change_type(Object *p_button, int p_index) {
 }
 
 void EditorPropertyDictionary::_add_key_value() {
-	// Do not allow nil as valid key. I experienced errors with this
-	if (object->get_new_item_key().get_type() == Variant::NIL) {
-		return;
+	// Treat empty object key as nil to prevent editor-only duplicates.
+	if (object->get_new_item_key().get_type() == Variant::OBJECT &&
+			object->get_new_item_key().is_null()) {
+		object->set_new_item_key(Variant());
```
I'm not sure about the reasoon behind this change. I fully agree that it is strange from the start but it is not linked in anyway to typed dictionarry and there is no consensus on how to treat it so it should be removed from this PR.

Repiteo replied:
I suppose this could be its own PR, but it's baffling that this exists at all. You can already add null as a key in scripts & with an empty object, it's specifically a non-object null that the *editor* can't add

ajreckof replied:
I know I was too starting to become mad because of it but I feel this should have its own PR to discuss wether we want to remove it entirely change it etc. Your input would really be appreciated here : #88797 to expose what you think should be the solution and why you think this can't create errors contrary to what states the commentary.

Repiteo replied:
Migrated relevant code to #89117

ajreckof commented on editor/editor_properties_array_dict.h:
```diff
@@ -174,22 +172,36 @@ class EditorPropertyDictionary : public EditorProperty {
+	void _page_changed(int p_page);
 
 protected:
+	Ref<EditorPropertyDictionaryObject> object;
+
+	bool updating = false;
+
 	static void _bind_methods();
 	void _notification(int p_what);
 
+	virtual void _edit_pressed();
+	virtual void _property_changed(const String &p_property, Variant p_value, const String &p_name = "", bool p_changing = false);
+	virtual void _change_type(Object *p_button, int p_index);
+	virtual void _change_type_menu(int p_index);
+
+	virtual void _add_key_value();
+	virtual void _object_id_selected(const StringName &p_property, ObjectID p_id);
+	virtual void _remove_pressed(int p_index);
+
```
why are all of those modifications needed?

Repiteo replied:
If you're talking about moving these functions from `private` to `protected`, then it's not strictly needed. Instead, it serves to mirror the current access levels of `EditorPropertyArray`. With how much else went into mirroring functionality & implementation, it wasn't too farfetched to adjust this as well

ajreckof replied:
most likely in this case it should be the other way around (ie. dictionary is right and it should be changed in Array) as it does not seem to have any reasons to be protected instead of private but also to be virtual I think.

Repiteo replied:
It's that way for array because the class is inherited by `EditorPropertyFontNamesArray`; now dictionary is setup to allow for similar inheritance if needed in the future

ajreckof replied:
So checked a bit more and it is right for array but should not be replicated in dictionary as there is no reasons to as there is no class inheriting from it (array as one which overrides those functions I assume) I would keep them as private.

ajreckof replied:
Oh I see now but then I think there is no reasons to bring that to EditorPropertyDictionary

Repiteo replied:
In terms of immediate use, it's true there's no real reason for it. However, as it doesn't end up making anything publically accessible that wasn't before, there isn't any real harm to it either. As there's already other function additions/changes in the class, I'm leaving this as-is

ajreckof commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -898,255 +972,45 @@ void EditorPropertyDictionary::update_property() {
-				default: {
-				}
+			bool is_untyped_dictionary = (i == amount ? key_type : value_type) == Variant::NIL;
+			Variant::Type type = value.get_type() == Variant::NIL && !is_untyped_dictionary
+					? i == amount ? key_type : value_type
+					: value.get_type();
```
why is there `value.get_type() == Variant::NIL`it seems it will prefer the value type instead of the dictionary type. it feels like it might create some bug and is one more check then needed.

Repiteo replied:
It's been a while so I don't entirely recall why that specific check is there. It very well could be a holdover from when `is_untyped_dictionary` wasn't yet a variable. Gonna have to test locally a bit to see if removing it impacts anything

Repiteo replied:
Yeah, looks like a holdover. Removed that & changed the name from "value" to "prop_val" to make it less ambiguous

ajreckof commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -898,255 +972,45 @@ void EditorPropertyDictionary::update_property() {
 				prop_name = "new_item_key";
 				value = object->get_new_item_key();
-			} else if (i == amount + 1) {
+				if (value.get_type() == Variant::NIL && key_type != Variant::NIL) {
+					VariantInternal::initialize(&value, key_type);
+					object->set_new_item_key(value);
+				}
+			} else {
 				prop_name = "new_item_value";
 				value = object->get_new_item_value();
+				if (value.get_type() == Variant::NIL && value_type != Variant::NIL) {
+					VariantInternal::initialize(&value, value_type);
+					object->set_new_item_value(value);
+				}
```
why are those only present for new_key/new_value?

Repiteo replied:
The values in the dictionary itself will have already been initalized by this point. Only the new key/value fields need to be initalized explicitly

ajreckof replied:
if it is about initialisation shouldn't it be in the initialisation function you created?

Repiteo replied:
That's what I assumed at first, but it turns out that would only initalize it for a single new key/value. As soon as they're added to the dictionary, the new key/value get reset back to a plain `Variant` & require initalization once more. Making the initalization happen in this check instead will guarantee that the editor property will always have an initalized new key/value to work with

ajreckof replied:
what you talk about seems like a regression from #88636. you might have wrongly merged it. Before the linked PR it was reverting them back too variant null but after this PR it should initialize them to the right type.

Repiteo replied:
I'm describing the behavior it had when this part of the PR was initially set up, so it being changed since then is entirely reasonable. I'll have to see if that reset behavior is still in place anywhere

Repiteo replied:
Yep, this was indeed fixed! The relevant code has been moved to `setup` instead

ajreckof commented on editor/editor_properties_array_dict.cpp:
```diff
@@ -1328,7 +1218,7 @@ void EditorPropertyLocalizableString::update_property() {
 
 	if (updated_val.get_type() == Variant::NIL) {
-		edit->set_text(TTR("Localizable String (Nil)")); // This provides symmetry with the array property.
+		edit->set_text(TTR("(Nil) Localizable String")); // This provides symmetry with the array property.
```
Not sure what this has to do with typed Dictionaries?

Repiteo replied:
Not strictly related, but it makes the "provides symmetry" comment actually true by making `(Nil)` preceed the text instead of following it. It felt weird having one of the explicitly implemented systems claiming to mirror each other not actually mirror correctly

Repiteo replied:
Reverted to keep focus strictly on typed dictionaries.

ajreckof reviewed (commented):
Only checked the part that I know best (ie `EditorPropertyDictionary`)

AThousandShips commented:
@lethefrost Please don't bump without contributing significant new information. Use the :+1: reaction button on the first post instead.

DaloLorn commented:
Looks like that last push broke most of the builds...

otonashixav commented:
The existing behaviour seems as good as it can be without introducing explicit co-/contra-variant types (either by `in out` notation or `ReadonlyDictionary`, `WriteonlyDictionary` etc.), which I don't expect soon. 

The untyped (or `Variant`-typed) `Array` and `Dictionary` are difficult to do anything about without breaking compatibility, however, 

* Ignoring those two, `Array[T]` and `Dictionary[K, V]`[^1] are invariant, that is, e.g. neither `Array[Resource]` nor `Array[Object]` is a subtype of the other, and, 
* As far as I'm aware, the engine does not contain methods or functions which mutate arrays and dictionaries passed as arguments.

Therefore, I think that as long as you never use the untyped variants in your code, you won't encounter issues caused by variance.

[^1]:  I hope this is the case, I don't intend to checkout the PR.

RadiantUwU commented:
```cpp
* thread #1, name = 'godot.linuxbsd.', stop reason = breakpoint 2.1
    frame #0: 0x000055555a46503d godot.linuxbsd.editor.dev.x86_64.llvm`TileSourceInspectorPlugin::can_handle(this=0x0000555576f145b0, p_object=0x0000000000000000) at tile_set_editor.cpp:1000:2
   997  }
   998 
   999  bool TileSourceInspectorPlugin::can_handle(Object *p_object) {
-> 1000         ERR_FAIL_NULL_V(p_object, false);
   1001         return p_object->is_class("TileSetAtlasSourceProxyObject") || p_object->is_class("TileSetScenesCollectionProxyObject");
   1002 }
   1003
(lldb) up
frame #1: 0x00005555594b8033 godot.linuxbsd.editor.dev.x86_64.llvm`EditorInspector::instantiate_property_editor(p_object=0x0000000000000000, p_type=INT, p_path=0x00007fffffff7588, p_hint=PROPERTY_HINT_NONE, p_hint_text=0x0000555561502cc8, p_usage=0, p_wide=false) at editor_inspector.cpp:2601:30
   2598
   2599 EditorProperty *EditorInspector::instantiate_property_editor(Object *p_object, const Variant::Type p_type, const String &p_path, PropertyHint p_hint, const String &p_hint_text, const uint32_t p_usage, const bool p_wide) {
   2600         for (int i = inspector_plugin_count - 1; i >= 0; i--) {
-> 2601                 if (!inspector_plugins[i]->can_handle(p_object)) {
   2602                         continue;
   2603                 }
   2604
(lldb) up
frame #2: 0x000055555971848d godot.linuxbsd.editor.dev.x86_64.llvm`EditorPropertyDictionary::update_property(this=0x0000555561502160) at editor_properties_array_dict.cpp:1160:17
   1157                                         new_prop = editor;
   1158                                 } else {
   1159                                         bool use_key = slot.index == EditorPropertyDictionaryObject::NEW_KEY_INDEX;
-> 1160                                         new_prop = EditorInspector::instantiate_property_editor(nullptr, value_type, "", use_key ? key_subtype_hint : value_subtype_hint,
   1161                                                         use_key ? key_subtype_hint_string : value_subtype_hint_string, PROPERTY_USAGE_NONE);
   1162                                 }
   1163                                 new_prop->set_selectable(false);
(lldb) up
frame #3: 0x00005555594c0176 godot.linuxbsd.editor.dev.x86_64.llvm`EditorInspector::update_tree(this=0x0000555560f47f30) at editor_inspector.cpp:3498:9
   3495                                 ep->set_doc_path(doc_path);
   3496                                 ep->set_internal(p.usage & PROPERTY_USAGE_INTERNAL);
   3497
-> 3498                                 ep->update_property();
   3499                                 ep->_update_pin_flags();
   3500                                 ep->update_editor_property_status();
   3501                                 ep->update_cache();
(lldb) up
frame #4: 0x00005555593d84f9 godot.linuxbsd.editor.dev.x86_64.llvm`EditorInspector::edit(this=0x0000555560f47f30, p_object=0x000055558146ff50) at editor_inspector.cpp:3595:3
   3592                         update_scroll_request = scroll_cache[object->get_instance_id()]; //done this way because wait until full size is accommodated
   3593                 }
   3594                 object->connect(CoreStringName(property_list_changed), callable_mp(this, &EditorInspector::_changed_callback));
-> 3595                 update_tree();
   3596         }
   3597
   3598         // Keep it available until the end so it works with both main and sub inspectors.
(lldb) up
frame #5: 0x000055555970c29c godot.linuxbsd.editor.dev.x86_64.llvm`EditorPropertyResource::update_property(this=0x0000555560f32f30) at editor_properties.cpp:3348:20
   3345                         }
   3346
   3347                         if (res.ptr() != sub_inspector->get_edited_object()) {
-> 3348                                 sub_inspector->edit(res.ptr());
   3349                                 _update_property_bg();
   3350                         }
   3351
(lldb) up
frame #6: 0x00005555594c0176 godot.linuxbsd.editor.dev.x86_64.llvm`EditorInspector::update_tree(this=0x000055556e27b470) at editor_inspector.cpp:3498:9
   3495                                 ep->set_doc_path(doc_path);
   3496                                 ep->set_internal(p.usage & PROPERTY_USAGE_INTERNAL);
   3497
-> 3498                                 ep->update_property();
   3499                                 ep->_update_pin_flags();
   3500                                 ep->update_editor_property_status();
   3501                                 ep->update_cache();
(lldb) up
frame #7: 0x00005555593d84f9 godot.linuxbsd.editor.dev.x86_64.llvm`EditorInspector::edit(this=0x000055556e27b470, p_object=0x00005555819cc9a0) at editor_inspector.cpp:3595:3
   3592                         update_scroll_request = scroll_cache[object->get_instance_id()]; //done this way because wait until full size is accommodated
   3593                 }
   3594                 object->connect(CoreStringName(property_list_changed), callable_mp(this, &EditorInspector::_changed_callback));
-> 3595                 update_tree();
   3596         }
   3597
   3598         // Keep it available until the end so it works with both main and sub inspectors.
(lldb) up
frame #8: 0x00005555596c50f2 godot.linuxbsd.editor.dev.x86_64.llvm`EditorNode::_edit_current(this=0x00005555674e0880, p_skip_foreign=false, p_skip_inspector_update=false) at editor_node.cpp:2490:45
   2487                 Node *current_node = Object::cast_to<Node>(current_obj);
   2488                 ERR_FAIL_NULL(current_node);
   2489
-> 2490                 InspectorDock::get_inspector_singleton()->edit(current_node);
   2491                 if (current_node->is_inside_tree()) {
   2492                         NodeDock::get_singleton()->set_node(current_node);
   2493                         SceneTreeDock::get_singleton()->set_selected(current_node);
   ```
   Crash occurs due to p_object being nullptr.

ChiliEater commented:
I'm currently trying out this branch with C# and I've run into an issue with enums. Using a type like `string` works fine:
```cs
[Export]
private Godot.Collections.Dictionary<string, Shader> Works = [];
```
![image](https://github.com/godotengine/godot/assets/32131318/5e8c160d-9a93-4cb1-83ec-125942e68a28)

Exporting just an enum also works fine:
```cs
enum Edition {
	Standard,
	Polychrome,
}

[Export]
private Edition EnumTest;
```
![image](https://github.com/godotengine/godot/assets/32131318/0674a9c3-fb06-48ea-af09-a23ce55c9184)

However, using that enum as a key results in some weirdness:
```cs
[Export]
private Godot.Collections.Dictionary<Edition, Shader> EditionShaders = [];
```
![image](https://github.com/godotengine/godot/assets/32131318/68f6e0a2-c49a-4317-aa78-77c1d57a543a)

As you can see, the dict shows up as having a Variant value type. Additionally, only the first enum actually shows up in the dropdown:
![image](https://github.com/godotengine/godot/assets/32131318/be7b6135-4fea-4da1-81d8-90d431860f00)

Is this a known issue or am I doing something wrong here?

Mickeon commented on doc/classes/Dictionary.xml:
```diff
@@ -284,6 +340,45 @@
+		<method name="is_typed" qualifiers="const">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if the dictionary is typed. Typed dictionaries can only store elements of their associated type and provide type safety for the [code][][/code] operator. Methods of typed dictionary still return [Variant].
```
This description is a bit misleading, given it only talks about the value, not the key.

Mickeon commented on doc/classes/Dictionary.xml:
```diff
@@ -202,6 +222,42 @@
+		<method name="get_typed_key_builtin" qualifiers="const">
+			<return type="int" />
+			<description>
+				Returns the [enum Variant.Type] constant for a typed dictionary key. If the [Dictionary] key is not typed, returns [constant TYPE_NIL].
```
```suggestion
				Returns the built-in [Variant] type of the typed dictionary's keys as a [enum Variant.Type] constant. If the keys are not typed, returns [constant TYPE_NIL]. See also [method is_typed_key].
```

Mickeon commented on doc/classes/Dictionary.xml:
```diff
@@ -202,6 +222,42 @@
+		<method name="get_typed_value_builtin" qualifiers="const">
+			<return type="int" />
+			<description>
+				Returns the [enum Variant.Type] constant for a typed dictionary value. If the [Dictionary] value is not typed, returns [constant TYPE_NIL].
```
```suggestion
				Returns the built-in [Variant] type of the typed dictionary's values as a [enum Variant.Type] constant. If the values are not typed, returns [constant TYPE_NIL]. See also [method is_typed_value].
```

Mickeon commented on doc/classes/Dictionary.xml:
```diff
@@ -202,6 +222,42 @@
+		<method name="get_typed_key_class_name" qualifiers="const">
+			<return type="StringName" />
+			<description>
+				Returns a class name of a typed [Dictionary] key of type [constant TYPE_OBJECT].
```
```suggestion
				Returns the [b]built-in[/b] class name of the typed dictionary's keys, if the built-in [Variant] type is [constant TYPE_OBJECT]. Otherwise, returns an empty [StringName]. See also [method is_typed_key] and [method Object.get_class].
```

Mickeon commented on doc/classes/Dictionary.xml:
```diff
@@ -202,6 +222,42 @@
+		<method name="get_typed_key_script" qualifiers="const">
+			<return type="Variant" />
+			<description>
+				Returns the script associated with a typed dictionary key tied to a class name.
```
```suggestion
				Returns the [Script] instance associated with this typed dictionary's keys, or [code]null[/code] if it does not exist. See also [method is_typed_key].
```

Mickeon commented on doc/classes/Dictionary.xml:
```diff
@@ -202,6 +222,42 @@
+		<method name="get_typed_value_script" qualifiers="const">
+			<return type="Variant" />
+			<description>
+				Returns the script associated with a typed dictionary value tied to a class name.
```
```suggestion
				Returns the [Script] instance associated with this typed dictionary's values, or [code]null[/code] if it does not exist. See also [method is_typed_value].
```

Mickeon commented on doc/classes/Dictionary.xml:
```diff
@@ -148,6 +148,19 @@
+			<param index="5" name="value_class_name" type="StringName" />
+			<param index="6" name="value_script" type="Variant" />
+			<description>
+				Creates a typed dictionary from the [param base] dictionary.
```
This is _way too succinct_, but I'm not sure how to nicely expand it in the same way that it is in Array. Maybe this'll do for now.
```suggestion
				Creates a typed dictionary from the [param base] dictionary. A typed dictionary can only contain keys and values of the given types, or that inherit from the given classes, as described by this constructor's parameters.
```

Mickeon reviewed (commented):
Some suggestions to make it consistent with my recent Array overhaul (https://github.com/godotengine/godot/pull/69451).

One thing I feel like could be done is refer to keys in plural ("_typed dictionary key_" -> "_typed dictionary's keys_"). This is to highlight that we are referring to the concept of key, not any single key present in the dictionary in question. Same goes for the values.

(Also please let this happen sooner than later. We know types are completely screwed up right now but the people have been yearning for this)

Repiteo commented:
@ChiliEater Good catch! Checked for myself & there's indeed some serious inconsistencies with how typed dictionaries are handling enums. This isn't a C#-thing either, as this happens in GDScript too. It's interesting though, because typed arrays seem to handle it completely fine; not sure where this discrepancy got introduced.

Repiteo commented:
Resolved! Turns out enums were also passing commas in their type hints, so the data was getting garbled as it expected the only comma to be separating key/value. Now dict types are separated by semicolon, which arguably kicks the can down the road but that whole stringhint system is due for an overhaul anyway.

In any case, the only weirdness I haven't resolved is something that's also affecting typed arrays: enums are listed as `int` in the inspector hint. More pertinently for dictionaries, this also applies to keys.

![24-07-09 18-13-32 godot windows editor dev x86_64](https://github.com/godotengine/godot/assets/17489292/15540867-1210-43ed-856d-e0986b577010)

ChiliEater commented:
I just got around to recompile with the new fix. It looks like something still isn't quite right. Since you mentioned that the problem occurred with both GDScript and C#, I tested both.

With GDScript it indeed works:
```gdscript
enum Bar {This, Is, Working}
@export var foo : Dictionary[Bar, Node]
```
![image](https://github.com/user-attachments/assets/5387459c-04cb-4fc2-9c4f-ef70d8f3e29e)

A similar C# snippet results in a mysterious extra entry to appear, as well as the value not being properly hinted:
```cs
enum Wowie { This, Is, Working, }
[Export]
private Godot.Collections.Dictionary<Wowie, Node> Scenes = [];
```
![image](https://github.com/user-attachments/assets/50982379-1629-41ec-ad84-584d3f691a30)

I should mention that I'm not 100% sure if I correctly updated Godot.NET.Sdk, so let me know if you can't reproduce this. Also, thanks for the work so far!

RobertoGongora commented:
> A similar C# snippet results in a mysterious extra entry to appear, as well as the value not being properly hinted:
> 
> ```cs
> enum Wowie { This, Is, Working, }
> [Export]
> private Godot.Collections.Dictionary<Wowie, Node> Scenes = [];
> ```

I'm not a C# developer so I'm sorry if this is not relevant, but that trailing comma after `Working` is not present in the GDScript version, so maybe the parser is grabbing that `,` and treating it as a new value, and since it's empty it outputs whatever random memory location the compiler is getting?

Again, no context whatsoever on how C# or GDScript's parser works, just throwing ideas :)

CsloudX commented:
Oh, like this feature.
IMO, i prefer it can define like this:
var dic : Dictionary{int, String}

it was vs Array:
var arr : Array[int]

Jordyfel commented:
I believe this needs the same treatment as typed arrays in #85024 - call `_find_resources()` on the type so it is recognized as an external resource in case it is a resource with a path, because that path will need to be remapped if the script is moved in the filesystem

Repiteo commented:
@ChiliEater Tried replicating this issue locally, and it took an embarassingly long time for me to realize that this got fixed after rebuilding the `.godot` folder. While that's not an _ideal_ solution, it at least reassures me that the PR itself is setup correctly!

![24-08-27 11-47-44 godot windows editor dev x86_64 mono](https://github.com/user-attachments/assets/15739554-4ed6-4a0d-8c09-40f3c1f0cba2)

reduz reviewed (approved):
PR looks very clean. Looks great on my end, but needs proper review on the GDScript side by GDScript maintainers.

nubels commented:
Look great! Does this support nested dictionaries like the one below, too? 
`
var inventory: Dictionary[String, Dictionary[String, int]] := {
	"food": {
		"apple": 10,
		"banana": 5,
	},
	"materials" {
		"copper": 5,
		"iron": 10,
	},
}
`

Mickeon commented:
I'm afraid not. That's a limitation of the current typing system that will eventually be tackled separately

DaloLorn commented:
You can probably work around that by making your inner dictionary a class that happens to contain a dictionary. Direct nesting of typed dictionaries seems a little overkill to me anyway... can't think of a use case.

(Not that Repiteo should go out of his way *not* to do it! Just doesn't seem like something worth holding this up over.)

Mickeon commented:
To be clear, you can nest typed dictionaries inside typed dictionaries, like you can nest typed arrays inside typed arrays, but GDScript won't know or care about it.

dalexeev commented on editor/connections_dialog.cpp:
```diff
@@ -576,6 +576,13 @@ String ConnectDialog::get_signature(const MethodInfo &p_method, PackedStringArra
 				}
 				break;
+			case Variant::DICTIONARY:
+				if (pi.hint == PROPERTY_HINT_DICTIONARY_TYPE && !pi.hint_string.is_empty() && !pi.hint_string.begins_with("res://")) {
+					type_name = "Dictionary[" + pi.hint_string + "]";
+				} else {
+					type_name = "Dictionary";
+				}
```
```suggestion
				type_name = "Dictionary";
				if (pi.hint == PROPERTY_HINT_DICTIONARY_TYPE && !pi.hint_string.is_empty()) {
					String key_hint = pi.hint_string.get_slice(";", 0);
					String value_hint = pi.hint_string.get_slice(";", 1);
					if (key_hint.is_empty() || key_hint.begins_with("res://")) {
						key_hint = "Variant";
					}
					if (value_hint.is_empty() || value_hint.begins_with("res://")) {
						value_hint = "Variant";
					}
					if (key_hint != "Variant" || value_hint != "Variant") {
						type_name = "Dictionary[" + key_hint + ", " + value_hint + "]";
					}
```

dalexeev commented on editor/editor_help.cpp:
<details>
<summary>Patch</summary>

```diff
diff --git a/editor/editor_help.cpp b/editor/editor_help.cpp
index 109819a353..1d928c6bd2 100644
--- a/editor/editor_help.cpp
+++ b/editor/editor_help.cpp
@@ -376,10 +376,10 @@ static void _add_type_to_rt(const String &p_type, const String &p_enum, bool p_i
 	}
 
 	p_rt->push_color(type_color);
-	bool add_array = false;
+	bool add_typed_container = false;
 	if (can_ref) {
 		if (link_t.ends_with("[]")) {
-			add_array = true;
+			add_typed_container = true;
 			link_t = link_t.trim_suffix("[]");
 			display_t = display_t.trim_suffix("[]");
 
@@ -388,22 +388,21 @@ static void _add_type_to_rt(const String &p_type, const String &p_enum, bool p_i
 			p_rt->pop(); // meta
 			p_rt->add_text("[");
 		} else if (link_t.begins_with("Dictionary[")) {
-			add_array = true;
+			add_typed_container = true;
 			link_t = link_t.trim_prefix("Dictionary[").trim_suffix("]");
 			display_t = display_t.trim_prefix("Dictionary[").trim_suffix("]");
-			PackedStringArray link_split = link_t.split(";");
-			PackedStringArray display_split = display_t.split(";");
-			link_t = link_split.get(1);
-			display_t = display_split.get(1);
 
-			p_rt->push_meta("#Dictionary"); // class
+			p_rt->push_meta("#Dictionary", RichTextLabel::META_UNDERLINE_ON_HOVER); // class
 			p_rt->add_text("Dictionary");
-			p_rt->pop();
+			p_rt->pop(); // meta
 			p_rt->add_text("[");
-			p_rt->push_meta("#" + link_split.get(0)); // class
-			p_rt->add_text(display_split.get(0));
-			p_rt->pop();
-			p_rt->add_text(",");
+			p_rt->push_meta("#" + link_t.get_slice(",", 0), RichTextLabel::META_UNDERLINE_ON_HOVER); // class
+			p_rt->add_text(_contextualize_class_specifier(display_t.get_slice(",", 0), p_class));
+			p_rt->pop(); // meta
+			p_rt->add_text(", ");
+
+			link_t = link_t.get_slice(",", 1);
+			display_t = _contextualize_class_specifier(display_t.get_slice(",", 1), p_class);
 		} else if (is_bitfield) {
 			p_rt->push_color(Color(type_color, 0.5));
 			p_rt->push_hint(TTR("This value is an integer composed as a bitmask of the following flags."));
@@ -422,7 +421,7 @@ static void _add_type_to_rt(const String &p_type, const String &p_enum, bool p_i
 	p_rt->add_text(display_t);
 	if (can_ref) {
 		p_rt->pop(); // meta
-		if (add_array) {
+		if (add_typed_container) {
 			p_rt->add_text("]");
 		} else if (is_bitfield) {
 			p_rt->push_color(Color(type_color, 0.5));
```

</details>

**Note:** Here `","` is used as a separator, not `;`.

Repiteo replied:
Is this assuming the below change of `"%s;%s"` to `"%s,%s"` is applied, or are you saying in this context it's already a comma regardless?

dalexeev replied:
There are several type hint formats:

1. GDScript type hint.
2. Doc type hint - how the type is stored in XML files. `DocTools` and `GDScriptDocGen` should write the hint in this format, `EditorHelp` and `make_rst.py` should read and display it in GDScript format.
3. `PROPERTY_HINT_DICTIONARY_TYPE` hint string.
4. `PROPERTY_HINT_TYPE_STRING` hint string (potentially).

dalexeev commented on core/variant/typed_dictionary.h:
```diff
@@ -0,0 +1,342 @@
+	static const Variant::Type VARIANT_TYPE = Variant::DICTIONARY;
+	static const GodotTypeInfo::Metadata METADATA = GodotTypeInfo::METADATA_NONE;
+	static inline PropertyInfo get_class_info() {
+		return PropertyInfo(Variant::DICTIONARY, String(), PROPERTY_HINT_DICTIONARY_TYPE, vformat("%s;%s", K::get_class_static(), V::get_class_static()));
```
```suggestion
		return PropertyInfo(Variant::DICTIONARY, String(), PROPERTY_HINT_DICTIONARY_TYPE, vformat("%s,%s", K::get_class_static(), V::get_class_static()));
```

Are there any considerations why this should be `;` and not `,`? It's less confusing than having both a comma and a semicolon as a separator in different places.

Repiteo replied:
Enums already use commas in their hint strings; opting for a semicolon prevents overlap between the two. It's fine to make the separator a comma once the enum data is parsed and discarded, when the representation is no longer a hint string

dalexeev replied:
There is a difference between `PROPERTY_HINT_ARRAY_TYPE` and `PROPERTY_HINT_TYPE_STRING`. This causes some confusion.

* The former uses the type name as a string and does not support specifying hint and hint string elements. Hint string examples:
  * `"int"` - `Array[int]`;
  * `"Node2D"` - `Array[Node2D]`;
  * `"Image"` - `Array[Image]`.
* The latter encodes the Variant type of an element, its hint and hint string using `:` and `/`. Hint string examples:
  * `"%d:" % [TYPE_INT]` - `Array[int]`;
  * `"%d/%d:%s" % [TYPE_OBJECT, PROPERTY_HINT_NODE_TYPE, "Node2D"]` - `Array[Node2D]`;
  * `"%d/%d:%s" % [TYPE_OBJECT, PROPERTY_HINT_RESOURCE_TYPE, "Image"]` - `Array[Image]`.

`PROPERTY_HINT_TYPE_STRING` [allows](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#class-globalscope-constant-property-hint-type-string) you to encode nested array types, even though they are not yet supported in GDScript.

Accordingly, for dictionaries we should also have different contents for `PROPERTY_HINT_DICTIONARY_TYPE` and `PROPERTY_HINT_TYPE_STRING`. For example, to encode `Dictionary[TKey, TValue]` you can use:

* `PROPERTY_HINT_DICTIONARY_TYPE`:
  * `TKeyString,TValueString`
* `PROPERTY_HINT_TYPE_STRING`:
  * `{TKeyBuiltin:|TValueBuiltin:}`
  * `{TKeyBuiltin/TKeyHint:TKeyHintString|TValueBuiltin/TValueHint:TValueHintString}`

We also need to think about cases of an array nested in a dictionary, a dictionary nested in an array, etc.

Mickeon replied:
It's insane how much of a mess this whole system is but let's tap our ears for this minor release.

Repiteo replied:
I haven't messed with `PROPERTY_HINT_TYPE_STRING` at all, so that's something I'll have to dig into. Honestly, I wasn't even aware that it could be used for setting up typed arrays; I always thought it was for enum-specific contexts.

And yes, this entire system is in need of a rework. However, the objective of this PR has always been to setup typed dictionaries with the existing system, warts and all.

dalexeev replied:
Regarding `PROPERTY_HINT_TYPE_STRING`, I think it's not urgent, we can figure it out later. The only thing that should be decided in this PR is what separator to use: comma or semicolon.

Repiteo replied:
When actually displayed to the user, the separator will be `, `. If we're talking internally, I suppose we could make the separator something universal. I'm not married to any one format, as I only went with `;` for the sake of enum compatibility.

Repiteo replied:
Gave this some thought/tests, and this might just have to *stay* messy to some extent, at least if we're wanting to keep compatibility with enum string hints. For now though, the internal separator is `;`, with `, ` being used for the GDScript type hint & user-visible contexts. Not gonna fuss too much over the semantics, as a typesystem rework would break compatibility anyway.

dalexeev commented on modules/gdscript/editor/gdscript_docgen.cpp:
```diff
@@ -84,6 +84,15 @@ void GDScriptDocGen::_doctype_from_gdtype(const GDType &p_gdtype, String &r_type
+				_doctype_from_gdtype(p_gdtype.get_container_element_type_or_variant(0), key, r_enum);
+				_doctype_from_gdtype(p_gdtype.get_container_element_type_or_variant(1), value, r_enum);
+				if (key != "Variant" && value != "Variant") {
+					r_type = "Dictionary[" + key + "," + value + "]";
```
Originally, the separator should be `;`, see the comments above.

dalexeev reviewed (commented):
Miscellaneous. I haven't started reviewing the GDScript part yet.

dalexeev reviewed (commented):
Part 2. I tested the PR, reviewed the unit tests, and applied some PRs that were included in GDScript after you started working on this. Still haven't reviewed the main GDScript changes, so there will probably be a Part 3. :)

Repiteo commented on core/variant/variant_parser.cpp:
Assuming you're talking about case 2: if both the key/value are untyped, the dictionary as a whole is untyped. This has the same affect as declaring an array `Array[Variant]`, where it's treated as functionally identical to the base version. Everything else mentioned is fixed, including changing `Nil` to `Variant`.

AThousandShips commented on modules/gdscript/tests/scripts/analyzer/features/typed_dictionary_usage.gd:
```diff
@@ -0,0 +1,214 @@
+	var two: Dictionary[int, int] = one
+
+	func check_passing() -> bool:
+		assert(str(one) == '{ 104: 401 }')
```
```suggestion
		Utils.check(str(one) == '{ 104: 401 }')
```
As per:
* https://github.com/godotengine/godot/pull/96229

AThousandShips commented on modules/gdscript/tests/scripts/runtime/features/typed_dictionary_init_with_untyped_in_literal.gd:
```diff
@@ -0,0 +1,7 @@
+func test():
+	var untyped: Variant = 32
+	var typed: Dictionary[int, int] = { untyped: untyped }
+	assert(typed.get_typed_key_builtin() == TYPE_INT)
```
Here as well

dalexeev commented on modules/gdscript/gdscript_disassembler.cpp:
```diff
@@ -978,6 +1079,12 @@ void GDScriptFunction::disassemble(const Vector<String> &p_code_lines) const {
 
 				incr += 5;
 			} break;
+			case OPCODE_RETURN_TYPED_DICTIONARY: {
+				text += "return typed dictionary ";
+				text += DADDR(1);
+
+				incr += 8;
+			} break;
```
Some typed array/dictionary opcodes don't have type disassembly. No need to fix now, just a note for future PR.

dalexeev commented on modules/gdscript/gdscript_function.h:
Please update `can_contain_object()`:

```c++
case Variant::DICTIONARY:
    if (has_container_element_types()) {
        return get_container_element_type_or_variant(0).can_contain_object() || get_container_element_type_or_variant(1).can_contain_object();
    }
    return true;
```

dalexeev commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -5247,6 +5455,62 @@ GDScriptParser::DataType GDScriptAnalyzer::type_from_property(const PropertyInfo
 			result.set_container_element_type(0, elem_type);
+		} else if (p_property.type == Variant::DICTIONARY && p_property.hint == PROPERTY_HINT_DICTIONARY_TYPE) {
+			// Check element type.
+			PackedStringArray elem_type_names = (p_property.hint_string).split(";");
```
```suggestion
			PackedStringArray elem_type_names = p_property.hint_string.split(";");
```

Repiteo replied:
Updated to instead utilize `get_slice` like the [above suggestion](https://github.com/godotengine/godot/pull/78656#discussion_r1738626887).

Repiteo commented on core/doc_data.cpp:
```diff
@@ -57,6 +59,8 @@ void DocData::return_doc_from_retinfo(DocData::MethodDoc &p_method, const Proper
 	} else if (p_retinfo.type == Variant::ARRAY && p_retinfo.hint == PROPERTY_HINT_ARRAY_TYPE) {
 		p_method.return_type = p_retinfo.hint_string + "[]";
+	} else if (p_retinfo.type == Variant::DICTIONARY && p_retinfo.hint == PROPERTY_HINT_DICTIONARY_TYPE) {
+		p_method.return_type = "Dictionary[" + p_retinfo.hint_string + "]";
```
Good call, this'll make `;` used *exclusively* in hint string contexts. It shouldn't overlap with enums either, as that's not currently supported in external bindings anyway.

dalexeev reviewed (approved):
Overall it looks good to me, the implementation is consistent with typed arrays. We debated for a long time whether we should resolve the [typed array issues](https://github.com/godotengine/godot-proposals/discussions/7364) first, but then we came to the conclusion that waiting would take too long. So if we have to break typed collection compatibility in the future, it will be both arrays and dictionaries.

In my opinion, this PR is already in a very good state to be merged. However, it's worth waiting for approval from @vnen as the main GDScript maintainer.

Of course, there are some things we might have missed or overlooked, but I think we could address them in follow-up PRs.

1. Attribute access does not have static checks, unlike index access. See the review comments.
2. Add support for binary serialization of typed dictionaries, like in #78219. Since this is a sensitive part, it's good to do in a separate PR.
3. `PROPERTY_HINT_DICTIONARY_TYPE` vs `PROPERTY_HINT_TYPE_STRING`. These hints have different formats and are sometimes mixed up in the engine.
4. The inspector part can be improved in the future. Currently, keys are displayed as string representations, which is not very convenient for viewing and editing.
5. Perhaps it makes sense to adjust the documentation, GDScript tests, disassembler, etc.

Thanks for your great work, @Repiteo!

---

<details>
<summary>Patch</summary>

```diff
diff --git a/modules/gdscript/tests/scripts/utils.notest.gd b/modules/gdscript/tests/scripts/utils.notest.gd
index 5d615d8557..1e2788f765 100644
--- a/modules/gdscript/tests/scripts/utils.notest.gd
+++ b/modules/gdscript/tests/scripts/utils.notest.gd
@@ -24,6 +24,11 @@ static func get_type(property: Dictionary, is_return: bool = false) -> String:
 				if str(property.hint_string).is_empty():
 					return "Array[<unknown type>]"
 				return "Array[%s]" % property.hint_string
+		TYPE_DICTIONARY:
+			if property.hint == PROPERTY_HINT_DICTIONARY_TYPE:
+				if str(property.hint_string).is_empty():
+					return "Dictionary[<unknown type>, <unknown type>]"
+				return "Dictionary[%s]" % str(property.hint_string).replace(";", ", ")
 		TYPE_OBJECT:
 			if not str(property.class_name).is_empty():
 				return property.class_name
@@ -188,6 +193,8 @@ static func get_property_hint_name(hint: PropertyHint) -> String:
 			return "PROPERTY_HINT_INT_IS_POINTER"
 		PROPERTY_HINT_ARRAY_TYPE:
 			return "PROPERTY_HINT_ARRAY_TYPE"
+		PROPERTY_HINT_DICTIONARY_TYPE:
+			return "PROPERTY_HINT_DICTIONARY_TYPE"
 		PROPERTY_HINT_LOCALE_ID:
 			return "PROPERTY_HINT_LOCALE_ID"
 		PROPERTY_HINT_LOCALIZABLE_STRING:
```

</details>

Repiteo commented on modules/gdscript/gdscript_analyzer.cpp:
Can you clarify the expected behavior here? Should all cases be causing a compile-time error?

dalexeev replied:
Yes, there should be no difference between `dict.x` and `dict["x"]` (or `dict[&"x"]`).

Repiteo replied:
New output:
```gd
var dict: Dictionary[int, int]

# Compile-time error: Cannot find member "x" in base "Dictionary[int, int]".
print(dict.x)
# Compile-time error: Invalid index type "String" for a base of type "Dictionary[int, int]".
print(dict["x"])

# Compile-time error: Cannot find member "x" in base "Dictionary[int, int]".
dict.x = 1
# Compile-time error: Invalid index type "String" for a base of type "Dictionary[int, int]".
dict["x"] = 1
```

Repiteo replied:
Oop, looks like some unit tests got borked from my fix; gonna see what's up with that

fire commented:
Some conflicts from upstream occurred.

AThousandShips commented:
Serialisation now makes `{}}` in some cases

AThousandShips commented on core/variant/variant_parser.cpp:
```diff
@@ -2036,40 +2176,109 @@ Error VariantWriter::write(const Variant &p_variant, StoreStringFunc p_store_str
 					}
 				}
 
 				p_store_string_func(p_store_string_ud, "}");
```
This should be moved up

AThousandShips commented:
This shows we need some unit tests for this as the unit tests completely missed this, only the doc generator showed it which is an issue

dalexeev commented on modules/gdscript/tests/scripts/utils.notest.gd:
```diff
@@ -24,6 +24,12 @@ static func get_type(property: Dictionary, is_return: bool = false) -> String:
+			if property.hint == PROPERTY_HINT_DICTIONARY_TYPE:
+				if str(property.hint_string).is_empty():
+					return "Dictionary[<unknown type>, <unknown type>]"
+				var split = str(property.hint_string).split(";")
+				return "Dictionary[%s, %s]" % [split[0] if split[0] else "Variant", split[1] if split[1] else "Variant"]
```
I would prefer to fix `hint_string` in `GDScriptParser::DataType::to_property_info()` and return to the previous version.

```suggestion
				return "Dictionary[%s]" % str(property.hint_string).replace(";", ", ")
```

Or at least use `String.get_slice()`:

```suggestion
				var key_type: String = str(property.hint_string).get_slice(";", 0)
				var value_type: String = str(property.hint_string).get_slice(";", 1)
				if key_type.is_empty():
					key_type = "Variant"
				if value_type.is_empty():
					value_type = "Variant"
				return "Dictionary[%s, %s]" % [key_type, value_type]
```

Repiteo commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -4371,6 +4371,14 @@ bool GDScriptParser::export_annotations(AnnotationNode *p_annotation, Node *p_ta
 		export_type.type_source = variable->datatype.type_source;
 	}
 
+	bool is_dict = false;
```
I can't speak for the other types of export annotations, but this is still needed for the basic `@export` scenario; otherwise, all exported dictionaries would show up as untyped in the inspector

Repiteo commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -4459,6 +4472,120 @@ bool GDScriptParser::export_annotations(AnnotationNode *p_annotation, Node *p_ta
 			return false;
 		}
+
+		if (is_dict) {
```
I thiiink this is fine. This bit was tricky because I had to parse key & value independently, so there was a fair bit of repeat code as a result. It does the same logic as the outer `if (is_array)` check, albeit resolved earlier.

vnen replied:
I think this is fine. But eventually we should extract the logic to avoid the repetition and make sure it's consistent.

vnen commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -5072,6 +5230,49 @@ Array GDScriptAnalyzer::make_array_from_element_datatype(const GDScriptParser::D
+		Ref<Script> script_type = p_key_element_datatype.script_type;
+		if (p_key_element_datatype.kind == GDScriptParser::DataType::CLASS && script_type.is_null()) {
+			Error err = OK;
+			Ref<GDScript> scr = GDScriptCache::get_shallow_script(p_key_element_datatype.script_path, err);
```
Would be better to use `get_depended_shallow_script()` here because it creates a proper dependency for the parser.

vnen reviewed (approved):
Reviewed mostly the GDScript part and have some small notes but nothing blocking. Thanks a lot for working on this and sorry again for taking so long to review.

ajreckof reviewed (approved):
I checked the code modifications on `EditorPropertyDictionary` changes and tested it. This looks good.

akien-mga commented:
Thanks! Amazing work :tada: :1st_place_medal: 

This is one of the most awaited features for GDScript users, and thus a major highlight for 4.4!

lomeat commented:
Why did you regret TS method? Like `Dictionary<{ id: string }>`?
Or `Dictionary{ "id": String }`
It can be used as
```gdscript
var user: Dictionary{ "name": String, "age": int } = { "name": "Michael", "age": 20 }
```
The best way is:
```gdscript
type User = Dictionary{}
var user: User = {}
```

xorblo-doitus commented:
@lomeat 
> Why did you regret TS method? Like `Dictionary<{ id: string }>`? Or `Dictionary{ "id": String }` It can be used as
> 
> ```gdscript
> var user: Dictionary{ "name": String, "age": int } = { "name": "Michael", "age": 20 }
> ```

A class would better suit this purpose:
```gdscript
class User extends RefCounted:
	var name: String
	var age: int
```

lomeat commented:
@xorblo-doitus is it okay for godot put many files to autoload with class definitions? (for global access without scenes). Or i t can be a `types.gd` file with many classes for this story.
